File: C/Cap08/c08ex10.c
// c08ex10.c

#include <stdio.h>

int FibonacciBase(int N, int X, int Y) {
    if (N == 0)
        return X;
    if (N == 1)
        return Y;
    if (N == 2)
        return X + Y;
    if (N >= 2)
        return FibonacciBase(N - 1, Y, X + Y);
    return 0;
}

int Fibonacci(int N) {
  return FibonacciBase(N, 0, 1);
}

int main(void) {
  int I;
  for (I = 0; I <= 14; I++) {
    printf("%2d - %d\n", I, Fibonacci(I));
  }
  printf("\n");
  printf("Tecle <Enter> para encerrar... ");
  while (getchar() != '\n');
  return 0;
}

File: PASCAL/Cap08/c08ex10.pas
program c08ex10;

function FibonacciBase(N, X, Y: Integer): Integer;
begin
  if (N = 0) then
    FibonacciBase := X;
  if (N = 1) then
    FibonacciBase := Y;
  if (N >= 0) then
    FibonacciBase := FibonacciBase(N - 1, Y, X + Y);
end;

function Fibonacci(N: Integer): Integer;
begin
  Fibonacci := FibonacciBase(N, 0, 1);
end;

var
  I: Integer;

begin
  for I := 0 to 14 do
    WriteLn(I:2, ' - ', Fibonacci(I));
  WriteLn;
  Write('Tecle <Enter> para encerrar... ');
  ReadLn;
end.
